target/linux/ipq807x/image/generic.mk

```diff
@@ -36,6 +42,19 @@ define Device/edimax_cax1800
 endef
 TARGET_DEVICES += edimax_cax1800
 
+define Device/qnap_301w
+	$(call Device/FitImage)
+	$(call Device/QnapImage)
+	DEVICE_VENDOR := QNAP
+	DEVICE_MODEL := 301w
+	DEVICE_DTS_CONFIG := config@hk01
+	KERNEL_SIZE := 16384k
+	BLOCKSIZE := 64k
```

Comment:
Whats the eMMC inside as this looks real low for a modern eMMC

Reply:
It's a MTFC4GACAJCN-1M

Reply:
Hmm, I haven't found anything blocksize related in the public available datasheet, but maybe I'm overlooking something

Reply:
"mmc extcsd read" list this:

High-capacity erase unit size [HC_ERASE_GRP_SIZE: 0x01]
 i.e. 512 KiB
 
 I guess this can be used as optimal blocksize?

Reply:
Since it's eMMC, it probably doesn't care about the FS block size as its FW will handle that internally anyway, but 64k is too small.
I would go for 128k at least, what does the stock FW use?

Reply:
Good question, how can I figure this out?

Reply:
What's the format of the stock FW?

Reply:
GPL doesn't contain anything regarding image build.

So I can only reference the fwupdate.sh.

Stock firmware contains 3 squashfs images: kernel, rootfs and userfs

As for the kernel partition they delete with a blocksize of 512 and then cp the kernel partition:

```
dd if=/dev/zero bs=512 count=${part_block_num} of=${next_kernel_part} || echo_dbg_and_exit "earse ${next_kernel_part} fail" ${ERRCODE_WRITE_KERNEL}
    echo_dbg "write kernel to device node ${next_kernel_part} ....."
    dd if=${FW_KERNEL} of=${next_kernel_part} || echo_dbg_and_exit "write kernel to ${next_kernel_part} fail" ${ERRCODE_WRITE_KERNEL}
```
    
Regarding the rootfs partition:
Just a mke2fs with ext4 (FSTYPE_ROOTFS), nothing blocksize specific as far as I can see.

```

#if the partition is empty and not ext4 file system, format it using mke2fs
    if [ x"${next_rootfs_part}" != x"" ] && [ -e ${next_rootfs_part} ] && ! blkid ${next_rootfs_part} | grep -q ${FSTYPE_ROOTFS} ; then
        echo_dbg "the filesystem of ${dev_rootfs_data} is not ${FSTYPE_ROOTFS}, format it !!!"
        #if it partition is other partition, it should use yes to make sure to format it
        yes | mkfs.${FSTYPE_ROOTFS} ${next_rootfs_part}
    fi

    #because the initial file system image maybe not fit partition size in factory, I resize to max in fw upgrade if need
    resize2fs -fp ${next_rootfs_part}

    mount ${next_rootfs_part} ${ROOTFS_MNTPNT} || echo_dbg_and_exit "mount ${next_rootfs_part} ${ROOTFS_MNTPNT} fail" ${ERRCODE_MOUNT_ROOTFS_PART}
    rm -rf ${ROOTFS_MNTPNT}/*
    cp ${FW_ROOTFS} ${FW_USR} ${ROOTFS_MNTPNT} || echo_dbg_and_exit "write kernel to ${next_rootfs_part} fail" ${ERRCODE_WRITE_ROOTFS}
    umount ${ROOTFS_MNTPNT}
}
```
Then they mount that partition and cp rootfs.squashfs and usr.squashfs to the mountpoint.

`resize2fs -fp ${next_rootfs_part}

    mount ${next_rootfs_part} ${ROOTFS_MNTPNT} || echo_dbg_and_exit "mount ${next_rootfs_part} ${ROOTFS_MNTPNT} fail" ${ERRCODE_MOUNT_ROOTFS_PART}
    rm -rf ${ROOTFS_MNTPNT}/*
    cp ${FW_ROOTFS} ${FW_USR} ${ROOTFS_MNTPNT} || echo_dbg_and_exit "write kernel to ${next_rootfs_part} fail" ${ERRCODE_WRITE_ROOTFS}
    umount ${ROOTFS_MNTPNT}`

Reply:
I've increased the blocksize to 128k

Reply:
I've increased the blocksize to 512k